config/management/operational/src/validator_state.rs

```diff
@@ -2,67 +2,123 @@
 }
 
 impl VerifyValidatorState {
     pub fn execute(self) -> Result<VerifyValidatorStateResult, Error> {
         // Load the config, storage backend and create a json rpc client.
         let config = self
-            .validator_config
-            .config()?
-            .override_json_server(&self.json_server);
+            .config
+            .load()?
+            .override_json_server(&self.json_server)
+            .override_validator_backend(&self.validator_backend.validator_backend)?;
         let storage = config.validator_backend();
         let client = JsonRpcClientWrapper::new(config.json_server);
         let owner_account = storage.account_address(OWNER_ACCOUNT)?;
 
-        // Verify if the validator is in the set
-        let in_validator_set = client
-            .validator_set(None)?
-            .iter()
-            .any(|vi| vi.account_address() == &owner_account);
+        let validator_infos = client.validator_set(None)?;
+
+        let mut result = VerifyValidatorStateResult {
+            consensus_key_match: None,
+            consensus_key_unique: None,
+            fullnode_network_key_match: None,
+            in_validator_set: None,
+            validator_network_key_match: None,
+        };
 
-        // TODO(khiemngo): consider return early if the validator is not in the set
+        // Verify if the validator is in the set.
+        // If it is not, return result and no need to do other checks.
+        result.in_validator_set = Some(
+            validator_infos
+                .iter()
+                .any(|vi| vi.account_address() == &owner_account),
+        );
+        if result.in_validator_set == Some(false) {
+            return Ok(result);
+        }
 
-        // Fetch the current on-chain config for this operator's owner
+        // Fetch the current on-chain config for this operator's owner.
+        // Check if the consensus key held in secure storage matches
+        // that registered on-chain.
         let validator_config = client.validator_config(owner_account).and_then(|vc| {
             DecryptedValidatorConfig::from_validator_config_resource(&vc, owner_account)
         })?;
-
         let storage_key = storage.ed25519_public_from_private(CONSENSUS_KEY)?;
-        let consensus_key_match = storage_key == validator_config.consensus_public_key;
+        result.consensus_key_match = Some(storage_key == validator_config.consensus_public_key);
+
+        // If the consensus key matches, check if it is unique
+        if result.consensus_key_match == Some(true) {
+            result.consensus_key_unique = Some(!validator_infos.iter().any(|vi| {
+                vi.account_address() != &owner_account && vi.consensus_public_key() == &storage_key
+            }));
+        }
+
+        // Check if the validator network key held in secure storage
+        // matches that registered on-chain.
+        let storage_key = storage.ed25519_public_from_private(VALIDATOR_NETWORK_KEY)?;
```

Comment:
Another key consistency check we should think about adding is that all keys (consensus, validator network and fullnode network) match locally when compared with the validator info of the validator set (i.e., the snapshot of the validator configs taken when a new epoch was started). We can do this in a future commit 😄

To elaborate, you already have the `validator_infos` here, which hold the snapshot of all validator configs when the epoch started. If we find that we are in the current validator set, we should then verify that the keys held in the validator set infos match ours locally. Does this make sense?

Reply:
Added `TODO` for this check.

I will follow up with you to ask some questions about this. Thanks!